File: 101. Symmetric Tree/101. Symmetric Tree.rb
def is_symmetric(root)
    return true if root.nil?
    is_mirror(root.left, root.right)
end

# Recursive approach
def is_mirror(left, right)
    return true if left.nil? && right.nil?
    return false if left.nil? || right.nil?
    
    left.val == right.val &&
    is_mirror(left.left, right.right) &&
    is_mirror(left.right, right.left)
end

# Iterative solution
def is_symmetric_iterative(root)
    return true if root.nil?
    
    queue = []
    queue.push(root.left)
    queue.push(root.right)
    
    while !queue.empty?
        left = queue.shift
        right = queue.shift
        
        next if left.nil? && right.nil?
        return false if left.nil? || right.nil?
        return false if left.val != right.val
        
        queue.push(left.left)
        queue.push(right.right)
        queue.push(left.right)
        queue.push(right.left)
    end
    true
end

File: 101. Symmetric Tree/101. Symmetric Tree.rs
use std::collections::VecDeque;
use std::cell::RefCell;
use std::rc::Rc;

impl Solution {
    // Recursive solution
    pub fn is_symmetric(root: Option<Rc<RefCell<TreeNode>>>) -> bool {
        match root {
            None => true,
            Some(node) => {
                let node = node.borrow();
                Self::is_mirror(&node.left, &node.right)
            }
        }
    }
    
    // Recursive approach
    fn is_mirror(left: &Option<Rc<RefCell<TreeNode>>>, right: &Option<Rc<RefCell<TreeNode>>>) -> bool {
        match (left, right) {
            (None, None) => true,
            (Some(l), Some(r)) => {
                let l = l.borrow();
                let r = r.borrow();
                l.val == r.val &&
                Self::is_mirror(&l.left, &r.right) &&
                Self::is_mirror(&l.right, &r.left)
            }
            _ => false
        }
    }
    
    // Iterative solution
    pub fn is_symmetric_iterative(root: Option<Rc<RefCell<TreeNode>>>) -> bool {
        match root {
            None => true,
            Some(node) => {
                let node = node.borrow();
                let mut queue = VecDeque::new();
                queue.push_back(node.left.clone());
                queue.push_back(node.right.clone());
                
                while !queue.is_empty() {
                    let left = queue.pop_front().unwrap();
                    let right = queue.pop_front().unwrap();
                    
                    match (left, right) {
                        (None, None) => continue,
                        (Some(l), Some(r)) => {
                            let l = l.borrow();
                            let r = r.borrow();
                            if l.val != r.val {
                                return false;
                            }
                            queue.push_back(l.left.clone());
                            queue.push_back(r.right.clone());
                            queue.push_back(l.right.clone());
                            queue.push_back(r.left.clone());
                        }
                        _ => return false
                    }
                }
                true
            }
        }
    }
}